File: leetcode/python/LEET134.py
from typing import *

class Solution:
    def canCompleteCircuit(self, gas: List[int], cost: List[int]) -> int:
        if sum(gas) < sum(cost):
            return -1

        start, fuel = 0, 0
        for i in range(len(gas)):

            if gas[i] + fuel < cost[i]:
                start = i + 1 # 0 ~ i까지는 출발점이 될 수 없다.
                fuel = 0
            else:
                fuel += gas[i] - cost[i]

        return start

sol = Solution()
print(sol.canCompleteCircuit(gas = [2,3,4], cost = [3,4,3]))

File: leetcode/java/LEET134.java

public class LEET134 {
    public static void main(String[] args) {
        System.out.println(canCompleteCircuit(new int[] {1,2,3,4,5}, new int[] {3,4,5,1,2}));
    }

    public static int canCompleteCircuit(int[] gas, int[] cost) {
        int sumGas = 0, sumCost = 0;
        int start = 0, fuel = 0;
        for (int i=0;i<gas.length;i++) {
            sumGas += gas[i];
            sumCost += cost[i];

            if (gas[i] + fuel < cost[i]) {
                start = i + 1;
                fuel = 0;
            } else {
                fuel += gas[i] - cost[i];
            }
        }
        return sumGas < sumCost ? -1 : start;
    }
}
